lib/block-supports/layout.php

```diff
@@ -557,47 +560,76 @@ function gutenberg_incremental_id_per_prefix( $prefix = '' ) {
+	$container_content_class   = wp_unique_id( 'wp-container-content-' );
+	$child_layout_declarations = array();
+	$child_layout_styles       = array();
 
-	if ( 'fixed' === $layout_from_parent || 'fill' === $layout_from_parent ) {
-		$container_content_class = wp_unique_id( 'wp-container-content-' );
+	$self_stretch = isset( $block['attrs']['style']['layout']['selfStretch'] ) ? $block['attrs']['style']['layout']['selfStretch'] : null;
 
-		$child_layout_styles = array();
+	if ( 'fixed' === $self_stretch && isset( $block['attrs']['style']['layout']['flexSize'] ) ) {
+		$child_layout_declarations['flex-basis'] = $block['attrs']['style']['layout']['flexSize'];
+		$child_layout_declarations['box-sizing'] = 'border-box';
+	} elseif ( 'fill' === $self_stretch ) {
+		$child_layout_declarations['flex-grow'] = '1';
+	}
 
-		if ( 'fixed' === $layout_from_parent && isset( $block['attrs']['style']['layout']['flexSize'] ) ) {
-			$child_layout_styles[] = array(
-				'selector'     => ".$container_content_class",
-				'declarations' => array(
-					'flex-basis' => $block['attrs']['style']['layout']['flexSize'],
-					'box-sizing' => 'border-box',
-				),
-			);
-		} elseif ( 'fill' === $layout_from_parent ) {
-			$child_layout_styles[] = array(
-				'selector'     => ".$container_content_class",
-				'declarations' => array(
-					'flex-grow' => '1',
-				),
-			);
-		}
+	if ( isset( $block['attrs']['style']['layout']['columnSpan'] ) ) {
+		$column_span                              = $block['attrs']['style']['layout']['columnSpan'];
+		$child_layout_declarations['grid-column'] = "span $column_span";
+	}
+	if ( isset( $block['attrs']['style']['layout']['rowSpan'] ) ) {
+		$row_span                              = $block['attrs']['style']['layout']['rowSpan'];
+		$child_layout_declarations['grid-row'] = "span $row_span";
+	}
+	$child_layout_styles[] = array(
+		'selector'     => ".$container_content_class",
+		'declarations' => $child_layout_declarations,
+	);
 
-		gutenberg_style_engine_get_stylesheet_from_css_rules(
-			$child_layout_styles,
-			array(
-				'context'  => 'block-supports',
-				'prettify' => false,
-			)
+	/**
+	 * If column span is set, it should be removed on small grids.
+	 */
+	if ( isset( $block['attrs']['style']['layout']['columnSpan'] ) && $block['attrs']['style']['layout']['parentColumnWidth'] ) {
+		$column_span_number  = floatval( $block['attrs']['style']['layout']['columnSpan'] );
+		$parent_column_width = $block['attrs']['style']['layout']['parentColumnWidth'];
+		$parent_column_value = floatval( $parent_column_width );
+		$parent_column_unit  = explode( $parent_column_value, $parent_column_width )[1];
+
+		// Adding 1 to the column span to make up for gap value.
+		$container_query_value = ( $column_span_number + 1 ) * $parent_column_value;
+		$container_query_value = $container_query_value . $parent_column_unit;
+
+		$child_layout_styles[] = array(
+			'query'        => "@container (max-width: $container_query_value )",
```

Comment:
This is similar to what we were testing all those moon ago for CSS layers:

https://github.com/WordPress/gutenberg/pull/41424/files#diff-b85df651c33208862897162a87698bad76f15535f6432b97b022a72e99230edeR112

Very nice to see that you've come up with a similar approach intuitively. 🍺  

I think it'd be worth digging deeper to see if we can use a container query or `"better_name"` to nest other CSS rules, e.g., in [CSS layers](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer) or [nesting](https://developer.mozilla.org/en-US/docs/Web/CSS/Nesting_selector) itself

I'd be very keen to help!

Reply:
> we were testing all those moons ago for CSS layers

To elaborate a little, much of the work relies on thinking of a way to structure the rules store, e.g., using arrays to  represent nested hierarchies.

Reply:
Ok ok I think I found the generic name for what we want: [at-rule](https://developer.mozilla.org/en-US/docs/Web/CSS/At-rule) ☺️ 

Regarding the storage structure, we can always change it later if it becomes evident that there's an advantage to representing nested hierarchies with arrays. The API wouldn't change either way; that's an implementation detail. The important thing now is making sure we have the right API.

Reply:
> Ok ok I think I found the generic name for what we want: at-rule ☺️

Yes! Love it.

> The important thing now is making sure we have the right API.

💯